cupyx/fallback_mode/utils.py

```diff
@@ -33,11 +42,115 @@ def _call_numpy(func, args, kwargs):
     Returns:
         Result after calling func.
     """
+    args = _convert_ndarrays(args)
 
     numpy_args, numpy_kwargs = data_transfer._get_numpy_args(args, kwargs)
 
     numpy_res = func(*numpy_args, **numpy_kwargs)
 
     cupy_res = data_transfer._get_cupy_result(numpy_res)
 
-    return cupy_res
+    return _get_fallback_ndarray(cupy_res)
+
+
+def _convert_ndarrays(args):
+    return tuple(
+        [i._array if isinstance(i, ndarray) else i for i in args])
+
+
+def _get_fallback_ndarray(res):
+
+    if isinstance(res, cp.ndarray):
+        return ndarray(res)
+
+    return res
+
+
+def make_method(name):
+    def method(self, *args, **kwargs):
+
+        args = _convert_ndarrays(args)
+
+        cupy_method = getattr(cp.ndarray, name)
+
+        res = cupy_method(self._array, *args, **kwargs)
+
+        return _get_fallback_ndarray(res)
+
+    return method
+
+
+def _create_magic_methods():
+
+    _magic_methods = [
+
+        # Comparison operators:
+        '__richcmp__',
+
+        # Truth value of an array (bool):
+        '__nonzero__',
+
+        # Unary operations:
+        '__neg__', '__pos__', '__abs__', '__invert__',
+
+        # Arithmetic:
+        '__add__', '__sub__', '__mul__', '__matmul__', '__div__',
+        '__truediv__', '__floordiv__', '__mod__', '__divmod__', '__pow__',
+        '__lshift__', '__rshift__', '__and__', '__or__', '__xor__',
+
+        # Arithmetic, in-place:
+        '__iadd__', '__isub__', '__imul__', '__idiv__', '__itruediv__',
+        '__ifloordiv__', '__imod__', '__ipow__', '__ilshift__', '__irshift__',
+        '__iand__', '__ior__', '__ixor__',
+
+        # For standard library functions:
+        '__copy__', '__deepcopy__', '__reduce__',
+
+        # Container customization:
+        '__iter__', '__len__', '__getitem__', '__setitem__',
+
+        # Conversion:
+        '__int__', '__float__', '__complex__', '__oct__', '__hex__',
+
+        # String representations:
+        '__repr__', '__str__'
+    ]
+
+    for method in _magic_methods:
+        setattr(ndarray, method, make_method(method))
+
+
+class ndarray:
+
+    def __init__(self, array):
```

Comment:
The arguments are different from `numpy.ndarray`.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html

If you want to initialize with CuPy's array, `ndarray.new()` should be used.

Reply:
Do you mean `ndarray.__new__`?

Reply:
Yes.

Reply:
There's no need for `ndarray` to support `cupy.ndarray.__init__`.
Actually, we've two different `ndarray` classes. `fallback_mode.utils.ndarray` and `fallback_mode.numpy.ndarray`. `fallback_mode.numpy.ndarray` will maintain the compatibility with `cupy.ndarray`. Whereas `utils.ndarray` is wrapper around it. `ndarray` created using `fallback_mode.numpy.ndarray` will be automatically converted to `utils.ndarray`.

Reply:
We don't want to expose `utils`. How about renaming `utils.py` to `_utils.py`?

Reply:
Okay